File: code_golf/C#/ExtractStringFromString.cs
/* 
 * 2/17/2021
 *
 * Extract a string from a given string
 *
 * code golf link:  https://codegolf.stackexchange.com/questions/50379/extract-a-string-from-a-given-string?rq=1
 *
 * You are given a string and two characters.You have to print the string between 
 * these characters from the string.
 * Input will first contain a string (not empty or null). In the next line, there
 * will be two characters separated by a space.
 */
using System;

namespace ExtractStringFromString
{
    class Program
    {
        static void Main(string[] args)
        {
            Console.Write("Enter string: ");
            var inputStr = Console.ReadLine();
            if (inputStr.Length > 100)
            {
                Console.WriteLine("Error: Input string greater than 100 characters.");
                return;
            }

            Console.Write("Enter endpoints: ");
            var endpoints = Console.ReadLine();
            if (endpoints.Length != 3 || endpoints[1] != ' ')
            {
                Console.WriteLine("Error: Endpoints in wrong format.");
                return;
            }
            if (endpoints[0] == endpoints[2])
            {
                Console.WriteLine("\"null\"");
                return;
            }
            if ((int)endpoints[0] < 32 || (int)endpoints[0] > 126 || (int)endpoints[2] < 32 || (int)endpoints[2] > 126)
            {
                Console.WriteLine("Error: Incorrect character(s) for endpoints. The endpoints characters can only have characters in the ASCII table range from space to tilde.");
                return;
            }

            Console.WriteLine(ExtractString(inputStr, endpoints));
        }

        public static string ExtractString(string inputStr, string endpoints)
        {
            var endpoint1 = endpoints[0];
            var endpoint2 = endpoints[2];
            var endpoint1Position = 0;
            var endpoint1Counter = 0;
            var endpoint2Position = 0;
            var endpoint2Counter = 0;
            for (int i = 0; i < inputStr.Length; i++)
            {
                if ((int)inputStr[i] < 32 || (int)inputStr[i] > 126 || (int)inputStr[i] < 32 || (int)inputStr[i] > 126)
                {
                    return "Error: Incorrect character(s) for input string. The input string can only have characters in the ASCII table range from space to tilde.";
                }
                if (inputStr[i] == endpoint1)
                {
                    endpoint1Counter++;
                    if (endpoint1Counter > 1)
                    {
                        return "\"null\"";
                    }
                    endpoint1Position = i;
                }
                if (inputStr[i] == endpoint2)
                {
                    endpoint2Counter++;
                    if (endpoint2Counter > 1)
                    {
                        return "\"null\"";
                    }
                    endpoint2Position = i;
                }
            }
            if (endpoint1Counter == 0 || endpoint2Counter == 0)
            {
                return "\"null\"";
            }

            if (endpoint1Position < endpoint2Position)
            {
                return '"' + inputStr.Substring(endpoint1Position + 1, endpoint2Position - endpoint1Position - 1) + '"';
            }
            else
            {
                return '"' + inputStr.Substring(endpoint2Position + 1, endpoint1Position - endpoint2Position - 1) + '"';
            }
        }
    }

}

File: code_golf/php/extract_string_from_string.php
<?php
/* 7/15/2021
 * Extract string from a given string
 * code golf link:  https://codegolf.stackexchange.com/questions/50379/extract-a-string-from-a-given-string?rq=1
 * 
 * You are given a string and two characters. You have to print the string between 
 * these characters from the string.
 * Input will first contain a string (not empty or null). In the next line, there 
 * will be two characters separated by a space.
 */

$inputStrings = [
    ["Hello World!", "H !"],
    ['<>code<>', "> <"],
    ['>code<', "> <"],
    ["What's what?", "' '"],
    ["abcdefghijklmnopqrstuvwxyz", "n z"],
    ["abcdefghijklmnopqrstuvwxyz", "z n"],
    ["Testing...", "e T"],
    ["agdagaga", "a"],
    ["Last test-case", "  -"],
];
echo '<pre>';
print_r($inputStrings);
echo '</pre>';

for ($i = 0; $i < count($inputStrings); $i++) {
    //echo "<br>";

    echo $inputStrings[$i][0] . "<br>";
    echo $inputStrings[$i][1] . "<br>";
    //echo $str . "<br>";

    $innerStr = getInnerString($inputStrings[$i][0], $inputStrings[$i][1]);

    echo "$innerStr<br><br>";
    //echo "<br>";
}

function getInnerString($str, $endpoints) {
    //echo "$str: {$str} <br>";
    //echo "\$endpoints: $endpoints<br>";
    //echo "endpoints[0]: $endpoints[0]<br>";
    //echo "endpoints[2]: $endpoints[2]<br>";

    if (strlen($str) < 3 || strlen($str) > 100) {
        return "Incorrect Input: Input string must be between 3 and 100 characters.<br>";
    }

    if (hasInvalidChars($str) || hasInvalidChars($endpoints)) {
        return "Incorrect Input: Input string and endpoints string can only have characters between a space character (32) and a tilde character '~' (127);<br>";
    }

    if ($endpoints && (strlen($endpoints) == 3) && ($endpoints[1] == " ") && ($endpoints[0] !== $endpoints[2])) {
        $endpoint1 = $endpoints[0];
        $endpoint2 = $endpoints[2];
    } else {
        return "Incorrect Input: Endpoints must be 2 different characters with a space between them.<br>";
    }

    $endpoint1Position = strpos($str, $endpoint1);
    //echo "position1: $endpoint1Position<br>";

    $endpoint2Position = strpos($str, $endpoint2);
    //echo "position2: $endpoint2Position<br>";

    if (isEndpointDuplicated($str, $endpoint1) || isEndpointDuplicated($str, $endpoint2)) {
        //echo '"null"' . "<br>";
        return '"null"';
    }

    if ($endpoint1Position < $endpoint2Position) {
        $innerStr = substr($str, $endpoint1Position + 1, $endpoint2Position - $endpoint1Position - 1);
    } else {
        $innerStr = substr($str, $endpoint2Position + 1, $endpoint1Position - $endpoint2Position - 1);
    }
    return '"' . $innerStr . '"';
}
 
function isEndpointDuplicated($str, $endpoint) {
    $charCount = 0;
    for ($i = 0; $i < strlen($str); $i++) {
        if ($str[$i] == $endpoint) {
            $charCount++;
        }
        if ($charCount > 1) {
            return true;
        }
    }
    return false;
}

// Check if input string has any characters outside of the correct ASCII range (space (32) to tilde (126)).
function hasInvalidChars($str) {
    for ($i = 0; $i < strlen($str); $i++) {
        if (ord($str[$i]) < 32 || ord($str[$i]) > 126) {
            return true;
        }
    }
    return false;
}